### DISCUSSION TITLE
Deploying/Publish with ClickOnce Guide

### DISCUSSION THREAD

**Discussion by amaitland** [cefsharp/CefSharp#4056] [General]
Looking for some help? Get priority support by [sponsoring the project](https://github.com/sponsors/amaitland)

---
[Publishing with ClickOnce](https://docs.microsoft.com/en-us/visualstudio/deployment/quickstart-deploy-using-clickonce-folder?view=vs-2022#publishing-with-clickonce) may require additional steps depending on your project format and structure.

- Currently works when `PlatformTarget` is `x86/x64`
- `AnyCPU` requires that you `SelfHost` the `BrowserSubProcess` see section below.

---
### Packages.config with a single project (csproj/vbproj)

If you are using the older Nuget [packages.config](https://docs.microsoft.com/en-us/nuget/reference/packages-config) format then you have two options

**Option 1:**
Add the following to a `PropertyGroup` in your `csproj/vbproj` file before the `.targets` entries (which are at the bottom).
```xml
<CefSharpBuildAction>Content</CefSharpBuildAction>
```

**Option2:**

Follow the `Microsoft` documentation to [Migrate from packages.config to PackageReference](https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference).

---
### Packages.config with multiple projects (csproj/vbproj)

If you have `CefSharp` installed in a library project that uses `pacakges.config` things become a little tricky. The [None](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#none) entries will copy the `CEF` resources to a directly referring project, changing to [Content](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#content) (changing the `CefSharpBuildAction` to `Content` changes to using `<Content/>` entries) means the files are only copied to the library project, not any referring projects.

1. Installing the `CefSharp` packages in your main project and adding `<CefSharpBuildAction>Content</CefSharpBuildAction>` to the first `PropertyGroup` of your main project file is one option as listed above.
2. Converting your library project to `PackageReference` should allow for the references to flow through to consuming projects as [buildTransitive](https://github.com/NuGet/Home/wiki/Allow-package--authors-to-define-build-assets-transitive-behavior). Noting that I haven't tested this with a mix of `PackageReference` and `packages.config`, so if you mix and match then option `1.` is the fall-back if this doesn't work.

---
### AnyCPU
Publishing when targeting `AnyCPU` behaves strangely, the x64 folder is missing a copy of CefSharp.dll which is required by the BrowserSubprocess. The file is copied correctly when the projects builds normally, so there's some quirk of the Publish process. I have yet to identify the root cause of the issue.

To workaround can self host the BrowserSubProcess to workaround the problem. Self Hosting means that instead of using `CefSharp.BrowserSubProcess.exe` to run the `renderer`, `gpu`, `network`, `audio` processes that are typically spawned, you use your own applications executable. You also **MUST** set performDependencyCheck to **false**.

#### WinForms
For `WinForms` your `Program.cs` would look something like:
```c#
public static class Program
{
	[STAThread]
	public static int Main(string[] args)
	{

		CefRuntime.SubscribeAnyCpuAssemblyResolver();

		//For Windows 7 and above, best to include relevant app.manifest entries as well
		Cef.EnableHighDPISupport();

		var exitCode = CefSharp.BrowserSubprocess.SelfHost.Main(args);

		if (exitCode >= 0)
		{
			return exitCode;
		}

		var settings = new CefSettings()
		{
			//By default CefSharp will use an in-memory cache, you need to specify a Cache Folder to persist data
			CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache"),
			BrowserSubprocessPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName
		};

		//Perform dependency check to make sure all relevant resources are in our output directory.
		//IMPORTANT: MUST set performDependencyCheck false
		Cef.Initialize(settings, performDependencyCheck: false, browserProcessHandler: null);

		var browser = new BrowserForm();
		Application.Run(browser);

		return 0;
	}
}
```
#### WPF
For `WPF` this is a little more difficult as the compiler generates the **Main** method for you. You need to:
- Create your own custom entry point (typically adding a `Program.cs` file with a `static Main` method would be the convention)
- Change the `Startup Object` to use your own main method, (Right click Properties on your project, Startup Object is a dropdown).

An example of  `Program.cs` for `WPF` might look like this:
```c#
public static class Program
{
	[STAThread]
	public static int Main(string[] args)
	{

		CefRuntime.SubscribeAnyCpuAssemblyResolver();

		//For Windows 7 and above, best to include relevant app.manifest entries as well
		Cef.EnableHighDPISupport();

		var exitCode = CefSharp.BrowserSubprocess.SelfHost.Main(args);

		if (exitCode >= 0)
		{
			return exitCode;
		}

		var settings = new CefSettings()
		{
			//By default CefSharp will use an in-memory cache, you need to specify a Cache Folder to persist data
			CachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "CefSharp\\Cache"),
			BrowserSubprocessPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName
		};

		//Perform dependency check to make sure all relevant resources are in our output directory.
		//IMPORTANT: MUST set performDependencyCheck false
		Cef.Initialize(settings, performDependencyCheck: false, browserProcessHandler: null);

		var app = new App();
		app.InitializeComponent();
		return app.Run();
	}
}
```

---
### Troubleshooting
Add the following to your `csproj/vbproj` file at the end

```xml
<Target Name="CefSharpAfterBuildDebug" AfterTargets="AfterBuild">
    <CallTarget Targets="CefSharpAfterBuildDiagnostic" />
</Target>
```

Change your `MSBuild project output verbosity` to `Normal`:
```
1. Go to Tools->Options menu in VS
2. Open Projects and Solutions->Build and Run
3. Change the value of the MSBuild project build output verbosity. Pick Normal or higher
```

After you build, check the output window for a block of text that looks like:
```
>Target CefSharpAfterBuildDebug:
>  Target CefSharpAfterBuildDiagnostic:
>    CefSharp After Build Diagnostic
>    CefSharpBuildAction = Content
>    CefSharpPropertiesLoaded = true
>    CefSharpTargetDir =
>    CefSharpTargetDirAnyCpu32 = x86\
>    CefSharpTargetDirAnyCpu64 = x64\
>    RuntimeIdentifier = win7-x64
>    EffectivePlatform =
>    Platform = x64
>    PlatformName = x64
>    Platforms = x86;x64;AnyCPU
>    PlatformTarget = x64
>    CefSharpPlatformTarget = x64
>    PlatformTargetAsMSBuildArchitecture = x64
>    TargetFramework = net472
>    TargetFrameworkVersion = v4.7.2
>    libcef.dll exists = bin.net472\x64\Debug\net472\libcef.dll
>    CefSharpTransformXmlDllPath = C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Web\Microsoft.Web.Publishing.Tasks.dll
>    NuGetProjectStyle = PackageReference
>    NuGetToolVersion = 5.11.1
>    OutputType = WinExe
>    OutDir = bin.net472\x64\Debug\net472\
>    Prefer32Bit = false
```

You can check `CefSharpBuildAction` was set correctly, confirm the `PlatformTarget` is as expected. Post this information when asking for help.

---
### Testing with MinimalExample

The [MinimalExample](https://github.com/cefsharp/CefSharp.MinimalExample) can be deployed using `ClickOnce` and I suggest anyone having problems follow these instructions and get that working as a reference. Then you can compare your project to the example.

**Option 1:**
- Open CefSharp.MinimalExample.sln in VS2019
- Build
- Close Solution
- Edit `CefSharp.MinimalExample.WinForms.csproj` adding `<CefSharpBuildAction>Content</CefSharpBuildAction>` to the first `PropertyGroup` (Example at [CefSharp.MinimalExample.WinForms.csproj#L31](https://github.com/cefsharp/CefSharp.MinimalExample/blob/cefsharp/91/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj#L31))
- Open CefSharp.MinimalExample.sln in VS2019
- Rebuild
- Comment out `Microsoft.Windows.Common-Controls` from `app.manifest` in `CefSharp.MinimalExample.WinForms`
- Right click `CefSharp.MinimalExample.WinForms` and publish
- Use default options
- Install ClickOnce deployed app.

Make sure that when installing on client machines that `Microsoft Visual C++ 2019 or higher` is installed.
You can use the `Windows Sandbox` for testing purposes, making sure to install  `Visual C++ 2019 or highter`
Loads as expected. 

---
A lot of work goes into maintaining supporting this project, if you are using `CefSharp` in a commercial application please support my efforts by [sponsoring the project](https://github.com/sponsors/amaitland).

**Comment by pamar1908** [comment]
In .Net Framework 4.8, it fails to load the "CefSharp.Core.Runtime.dll" even if we specify the CefSharpBuildAction as in option 1.

**Reply by amaitland** [reply]
Start your own discussion and answer the questions at https://github.com/cefsharp/CefSharp/wiki/Questions-&-Support

https://github.com/cefsharp/CefSharp/discussions/new/choose
